### PR TITLE
Make startswith work with IO objects

### LINKED ISSUE #40616: Make startswith work with IO objects
This would be as useful as `peek` and other functions. A simple implementation (for strings) could be

```julia
"""
    startswith(io::IO, prefix::AbstractString)

Check if an `IO` object starts with a prefix.
"""
function startswith(io::IO, prefix::AbstractString)
    pos = position(io)
    s = String(read(io, length(prefix)))
    seek(io, pos)
    return startswith(s, prefix)
end
```

### PR DESCRIPTION
**Pull request by ArunS-tack**
closes #40616.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
I suppose this should use `mark` and `reset`?

File: base/strings/util.jl [hunk 1]
```diff
@@ -57,6 +57,20 @@ function startswith(a::Union{String, SubString{String}},
...
+Check if an `IO` object starts with a prefix.
+"""
+function Base.startswith(io::IO, prefix::Union{AbstractString,Base.Chars})
+    pos = position(io)
+    s = _getminprefix(io, prefix)
+    seek(io, pos)
+    return startswith(s, prefix)
+end
+_getminprefix(io::IO, prefix::Union{AbstractString,AbstractChar}) = String(read(io, length(prefix)))
+_getminprefix(io::IO, chars::Union{Tuple{Vararg{<:AbstractChar}},AbstractVector{<:AbstractChar},Set{<:AbstractChar}}) = _getminprefix(io, first(chars))
```
Comment (JeffBezanson) [inline base/strings/util.jl:72]: These do not correctly handle unicode. If the prefix is a `Chars`, this should read one `Char` from the stream and see if it is in the set.  The prefix should also be converted to `String`, then we can use `ncodeunits` to know how many bytes to read.

**Comment by ArunSanganal** [comment]
> I suppose this should use `mark` and `reset`?

I am not quite familiar with those. Where should I add them?

File: base/strings/util.jl (hunk 1 above)
Comment (stevengj) [inline base/strings/util.jl:72]: something like:
```suggestion
function Base.startswith(io::IO, prefix::Base.Chars)
    mark(io)
    c = read(io, Char)
    reset(io)
    return c in prefix
end
function Base.startswith(io::IO, prefix::Union{String,SubString{String}})
    mark(io)
    s = read(io, ncodeunits(prefix))
    reset(io)
    return s == codeunits(prefix)
end
Base.startswith(io::IO, prefix::AbstractString) = startswith(io, String(prefix))
    pos = position(io)
    s = _getminprefix(io, prefix)
    seek(io, pos)
    return startswith(s, prefix)
end
_getminprefix(io::IO, prefix::Union{AbstractString,AbstractChar}) = String(read(io, length(prefix)))
_getminprefix(io::IO, chars::Union{Tuple{Vararg{<:AbstractChar}},AbstractVector{<:AbstractChar},Set{<:AbstractChar}}) = _getminprefix(io, first(chars))
```

File: test/strings/basic.jl [hunk 2]
```diff
@@ -164,6 +164,9 @@ end
     @test endswith(y)(y)
     @test endswith(z, z)
     @test endswith(z)(z)
+    #40616 startswith for IO objects
+    io = IOBuffer("JuliaLang")
+    @test startswith(io, "Julia")
```
Comment (stevengj) [inline test/strings/basic.jl:169]: Needs a test for the `Chars` method too.

Comment (stevengj) [reply test/strings/basic.jl:169]: Probably should be something like
```jl
let s = "JuliaLang", io = IOBuffer(s)
    for prefix in ("Julia", "July", "Ju", 'J', 'x', ('j','J'))
        @test startswith(io, prefix) == startswith(s, prefix)
    end
end
```

**Review by stevengj — DISMISSED** [review]

**Comment by stevengj** [comment]
Looks like it just needs a NEWS item, something like:
```
* `startswith` now supports seekable `IO` streams ([#43055]).
```

**Comment by KristofferC** [comment]
Personally, I think this is kind of a strange API for an `IO`. For example, in the implementation here you need to support `mark` and `reset` but those are not actually needed to check if an `IO` "starts with" something. In the same vein, there are many other "stringy" type of functions (`contains`, `endswith`, etc) that one could theoretically call on an `IO` because you can always materialize a String from an `IO´ and then reset.

But I think it is best to keep `IO` to be used for reading/writing and then you just compose that with the string functions to get the desired behavior. That avoids many questions like what state the IO object should be in after this function is called and makes it up to the user to decide on a case by case basis.

**Comment by stevengj** [comment]
As @schneiderfelipe pointed out in the original issue, it's a nice complement to `peek` — indeed, checking if the subsequent stream starts with something is the basic use case for `peek`, and this is both a generalization that and arguably nicer API for that kind of query.

**Review by stevengj — APPROVED** [review]

File: base/strings/util.jl [hunk 3]
```diff
@@ -67,6 +67,25 @@ function startswith(a::Union{String, SubString{String}},
     end
 end
 
+"""
+    startswith(io::IO, prefix::Union{AbstractString,Base.Chars})
+
+Check if an `IO` object starts with a prefix.
+"""
+function Base.startswith(io::IO, prefix::Base.Chars)
+    mark(io)
```
Comment (stevengj) [inline base/strings/util.jl:76]: Seems worth it to optimize the common case of an ASCII char:
```suggestion
    prefix isa AbstractChar && prefix ≤ '\x7f' && return peek(io) == prefix % UInt8
    mark(io)
```

File: base/strings/util.jl [hunk 4]
```diff
@@ -67,6 +67,25 @@ function startswith(a::Union{String, SubString{String}},
     end
 end
 
+"""
+    startswith(io::IO, prefix::Union{AbstractString,Base.Chars})
+
+Check if an `IO` object starts with a prefix.
```
Comment (stevengj) [inline base/strings/util.jl:73]:
```suggestion
Check if an `IO` object starts with a prefix.  See also [`peek`](@ref).
```

**Pull request merged**